pkg/util/token_bucket.go

```diff
@@ -0,0 +1,95 @@
+package util
+
+import (
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type TokenBucket struct {
```

Comment:
Public functions of the existing rate limiter seem quite different from mine, where I need `ForceRetrieve` and `Refund` as well.

Reply:
Isn't `Retrieve` and `AllowN` the same? Also isn't `ForceRetrieve` and `ReserveN` the same?

Reply:
Yes it looks like `AllowN` is same as `Retrieve`, but `ReserveN` returns a `Reservation` that indicates how long the caller must wait before n events happen.

I feel like token bucket code is simple enough, and I don't see big advantage of reusing rate_limiter for it (it also includes concepts like burst and reservation, which is unnecessary for the purpose of this token bucket)

Reply:
> ReserveN returns a Reservation that indicates how long the caller must wait before n events happen.

True. But if you don't wait and act immediately it'll behave just like `ForceRetrieve()`

Reply:
I think I'd still like to have `Refund` so that I can take out same amount of tokens from all buckets when it succeed (and take out 0 token from all buckets when one of the buckets fail)

Reply:
I'm okay with using a custom implementation.